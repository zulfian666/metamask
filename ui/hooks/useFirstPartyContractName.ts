import { NameType } from '@metamask/name-controller';
import { useSelector } from 'react-redux';
import { getCurrentChainId } from '../selectors';
import {
  EXPERIENCES_TYPE,
  FIRST_PARTY_CONTRACT_NAMES,
} from '../../shared/constants/first-party-contracts';

export type UseFirstPartyContractNameRequest = {
  value: string;
  type: NameType;
  variation?: string;
};

export function useFirstPartyContractNames(
  requests: UseFirstPartyContractNameRequest[],
): (string | null)[] {
  const currentChainId = useSelector(getCurrentChainId);

  return requests.map(({ type, value, variation }) => {
    if (type !== NameType.ETHEREUM_ADDRESS) {
      return null;
    }

    const chainId = variation ?? currentChainId;
    const normalizedValue = value.toLowerCase();

    return (
      Object.keys(FIRST_PARTY_CONTRACT_NAMES).find(
        (name) =>
          FIRST_PARTY_CONTRACT_NAMES[name as EXPERIENCES_TYPE]?.[
            chainId
          ]?.toLowerCase() === normalizedValue,
      ) ?? null
    );
  });
}

export function useFirstPartyContractName(
  value: string,
  type: NameType,
  variation?: string,
): string | null {
  return useFirstPartyContractNames([{ value, type, variation }])[0];
}
