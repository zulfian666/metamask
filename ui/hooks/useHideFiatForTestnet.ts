import { useSelector } from 'react-redux';
import type { Hex } from '@metamask/utils';
import { getShowFiatInTestnets, getCurrentChainId } from '../selectors';
import { TEST_NETWORK_IDS } from '../../shared/constants/network';

/**
 * Returns true if the fiat value should be hidden for testnet networks.
 *
 * @param providedChainId
 * @returns boolean
 */
export const useHideFiatForTestnet = (providedChainId?: Hex): boolean => {
  const showFiatInTestnets = useSelector(getShowFiatInTestnets);
  const currentChainId = useSelector(getCurrentChainId);
  const chainId = providedChainId ?? currentChainId;
  return TEST_NETWORK_IDS.includes(chainId) && !showFiatInTestnets;
};
