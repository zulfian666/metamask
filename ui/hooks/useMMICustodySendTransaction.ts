import { useDispatch, useSelector } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { TransactionMeta } from '@metamask/transaction-controller';
import { showCustodianDeepLink } from '@metamask-institutional/extension';
import { updateAndApproveTx } from '../store/actions';
import { AccountType, CustodyStatus } from '../../shared/constants/custody';
import { getMostRecentOverviewPage } from '../ducks/history/history';
import { clearConfirmTransaction } from '../ducks/confirm-transaction/confirm-transaction.duck';
import { getAccountType } from '../selectors/selectors';
import { mmiActionsFactory } from '../store/institutional/institution-background';
import { showCustodyConfirmLink } from '../store/institutional/institution-actions';

type MMITransactionMeta = TransactionMeta & {
  txParams: { from: string };
  custodyStatus: CustodyStatus;
  metadata: Record<string, unknown>;
};

export function useMMICustodySendTransaction() {
  const dispatch = useDispatch();
  const history = useHistory();
  const mmiActions = mmiActionsFactory();
  const accountType = useSelector(getAccountType);
  const mostRecentOverviewPage = useSelector(getMostRecentOverviewPage);

  const custodyTransactionFn = async (_transactionData: TransactionMeta) => {
    const confirmation = _transactionData as MMITransactionMeta;

    if (confirmation && accountType === AccountType.CUSTODY) {
      confirmation.custodyStatus = CustodyStatus.CREATED;
      confirmation.metadata = confirmation.metadata || {};

      dispatch(mmiActions.setWaitForConfirmDeepLinkDialog(true));

      const txId = confirmation.id;
      const fromAddress = confirmation.txParams.from;
      const closeNotification = false;

      await dispatch(updateAndApproveTx(confirmation, true, ''));
      showCustodianDeepLink({
        dispatch,
        mmiActions,
        txId,
        fromAddress,
        closeNotification,
        isSignature: false,
        custodyId: '',
        onDeepLinkFetched: () => undefined,
        onDeepLinkShown: () => {
          dispatch(clearConfirmTransaction());
        },
        showCustodyConfirmLink,
      });
    } else {
      // Non Custody accounts follow normal flow
      await dispatch(updateAndApproveTx(_transactionData, true, ''));
      dispatch(clearConfirmTransaction());
      history.push(mostRecentOverviewPage);
    }
  };

  return { custodyTransactionFn };
}
