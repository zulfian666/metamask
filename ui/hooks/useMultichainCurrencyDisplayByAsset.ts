import { BigNumber } from 'bignumber.js';
import { useMemo } from 'react';
import { formatCurrency } from '../helpers/utils/confirm-tx.util';
import {
  getMultichainCoinRates,
  getMultichainCurrentCurrency,
} from '../selectors/multichain';
import { NativeAsset } from '../components/multichain/asset-picker-amount/asset-picker-modal/types';
import { useMultichainSelector } from './useMultichainSelector';

export const useMultichainCurrencyDisplayByAsset = ({
  assetDetails,
  amount,
}: {
  assetDetails: NativeAsset & {
    balance: string;
    details: { decimals: number };
  };
  amount: string;
}): {
  feeInFiat: string;
  displayValueFee: string;
} => {
  const conversionRates = useMultichainSelector(getMultichainCoinRates);
  const currency = useMultichainSelector(getMultichainCurrentCurrency);

  const feeInFiat = useMemo(() => {
    if (!conversionRates || !amount) {
      return '';
    }

    // rates from the rates controller uses symbol as the key
    const conversionRate =
      conversionRates[assetDetails.symbol.toLowerCase()]?.conversionRate ?? '0';

    // rates are in the largest  denomination of the asset
    // e.g. 1 ETH = 1000000000000000000 wei
    // e.g. 1 BTC = 100000000 satoshis
    const fiatValue = new BigNumber(amount).mul(new BigNumber(conversionRate));
    const formattedValue = formatCurrency(fiatValue.toString(), currency);

    return formattedValue;
  }, [amount, conversionRates]);

  const displayValueFee = `${amount} ${assetDetails.symbol}`;

  return { feeInFiat, displayValueFee };
};
