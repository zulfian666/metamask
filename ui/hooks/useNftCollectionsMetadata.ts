import { useEffect, useMemo, useState } from 'react';
import { Collection } from '@metamask/assets-controllers';
import { useSelector } from 'react-redux';
import { TokenStandard } from '../../shared/constants/transaction';
import { getCurrentChainId } from '../selectors';
import {
  getNFTContractInfo,
  getTokenStandardAndDetails,
} from '../store/actions';

export type UseNftCollectionsMetadataRequest = {
  value: string;
};

type CollectionsData = {
  [key: string]: Collection;
};

// For now, we only support ERC721 tokens
const SUPPORTED_NFT_TOKEN_STANDARDS = [TokenStandard.ERC721];

export function useNftCollectionsMetadata(
  requests: UseNftCollectionsMetadataRequest[],
) {
  const [collectionsMetadata, setCollectionsMetadata] =
    useState<CollectionsData>({});
  const chainId = useSelector(getCurrentChainId);

  const memoisedContracts = useMemo(() => {
    return requests
      .filter(({ value }) => value)
      .map(({ value }) => value.toLowerCase());
  }, [requests]);

  useEffect(() => {
    const fetchCollections = async () => {
      try {
        const contractStandardsResponses = await Promise.all(
          memoisedContracts.map((contractAddress) =>
            getTokenStandardAndDetails(contractAddress, chainId),
          ),
        );

        const supportedNFTContracts = memoisedContracts.filter(
          (_contractAddress, index) =>
            SUPPORTED_NFT_TOKEN_STANDARDS.includes(
              contractStandardsResponses[index].standard as TokenStandard,
            ),
        );

        if (supportedNFTContracts.length === 0) {
          return;
        }

        const collectionsResult = await getNFTContractInfo(
          supportedNFTContracts,
          chainId,
        );

        const collectionsData: CollectionsData =
          collectionsResult.collections.reduce(
            (acc: CollectionsData, collection, index) => {
              acc[supportedNFTContracts[index]] = {
                name: collection?.name,
                image: collection?.image,
                isSpam: collection?.isSpam,
              };
              return acc;
            },
            {},
          );

        setCollectionsMetadata(collectionsData);
      } catch (error) {
        // Ignore the error due to API failure
      }
    };

    if (memoisedContracts.length > 0) {
      fetchCollections();
    }
  }, [memoisedContracts, chainId]);

  return collectionsMetadata;
}
