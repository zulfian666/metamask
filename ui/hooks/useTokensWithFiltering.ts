import { useCallback, useMemo } from 'react';
import { useSelector } from 'react-redux';
import { isEqual } from 'lodash';
import { ChainId, hexToBN } from '@metamask/controller-utils';
import { Hex } from '@metamask/utils';
import {
  getAllTokens,
  getCurrentCurrency,
  getSelectedInternalAccountWithBalance,
  getShouldHideZeroBalanceTokens,
  getTokenExchangeRates,
} from '../selectors';
import { getConversionRate } from '../ducks/metamask/metamask';
import {
  SWAPS_CHAINID_DEFAULT_TOKEN_MAP,
  SwapsTokenObject,
  TokenBucketPriority,
} from '../../shared/constants/swaps';
import { getValueFromWeiHex } from '../../shared/modules/conversion.utils';
import { EtherDenomination } from '../../shared/constants/common';
import {
  AssetWithDisplayData,
  ERC20Asset,
  NativeAsset,
  TokenWithBalance,
} from '../components/multichain/asset-picker-amount/asset-picker-modal/types';
import { AssetType } from '../../shared/constants/transaction';
import { isSwapsDefaultTokenSymbol } from '../../shared/modules/swaps.utils';
import { useTokenTracker } from './useTokenTracker';
import { getRenderableTokenData } from './useTokensToSearch';

/*
Sorts tokenList by query match, balance/popularity, all other tokens
*/
export const useTokensWithFiltering = (
  tokenList: Record<string, SwapsTokenObject>,
  topTokens: { address: string }[],
  chainId: ChainId | Hex,
  sortOrder: TokenBucketPriority = TokenBucketPriority.owned,
) => {
  // Only inlucdes non-native tokens
  const allDetectedTokens = useSelector(getAllTokens);
  const { address: selectedAddress, balance: balanceOnActiveChain } =
    useSelector(getSelectedInternalAccountWithBalance);

  const allDetectedTokensForChainAndAddress =
    allDetectedTokens?.[chainId]?.[selectedAddress] ?? [];

  const shouldHideZeroBalanceTokens = useSelector(
    getShouldHideZeroBalanceTokens,
  );
  const {
    tokensWithBalances: erc20TokensWithBalances,
  }: { tokensWithBalances: TokenWithBalance[] } = useTokenTracker({
    tokens: allDetectedTokensForChainAndAddress,
    address: selectedAddress,
    hideZeroBalanceTokens: Boolean(shouldHideZeroBalanceTokens),
  });

  const tokenConversionRates = useSelector(getTokenExchangeRates, isEqual);
  const conversionRate = useSelector(getConversionRate);
  const currentCurrency = useSelector(getCurrentCurrency);

  const sortedErc20TokensWithBalances = useMemo(
    () =>
      erc20TokensWithBalances.sort(
        (a, b) => Number(b.string) - Number(a.string),
      ),
    [erc20TokensWithBalances],
  );

  const filteredTokenListGenerator = useCallback(
    (shouldAddToken: (symbol: string, address?: string) => boolean) => {
      const buildTokenData = (
        token: SwapsTokenObject,
      ):
        | AssetWithDisplayData<NativeAsset>
        | AssetWithDisplayData<ERC20Asset>
        | undefined => {
        if (shouldAddToken(token.symbol, token.address)) {
          return getRenderableTokenData(
            {
              ...token,
              type: isSwapsDefaultTokenSymbol(token.symbol, chainId)
                ? AssetType.native
                : AssetType.token,
              image: token.iconUrl,
            },
            tokenConversionRates,
            conversionRate,
            currentCurrency,
            chainId,
            tokenList,
          );
        }
        return undefined;
      };

      return (function* (): Generator<
        AssetWithDisplayData<NativeAsset> | AssetWithDisplayData<ERC20Asset>
      > {
        const balance = hexToBN(balanceOnActiveChain);
        const nativeToken = buildTokenData({
          ...SWAPS_CHAINID_DEFAULT_TOKEN_MAP[
            chainId as keyof typeof SWAPS_CHAINID_DEFAULT_TOKEN_MAP
          ],
          ...(sortOrder === TokenBucketPriority.owned
            ? {
                balance: balanceOnActiveChain,
                string: getValueFromWeiHex({
                  value: balance,
                  numberOfDecimals: 4,
                  toDenomination: EtherDenomination.ETH,
                }),
              }
            : {}),
        });
        if (nativeToken) {
          yield nativeToken;
        }

        if (sortOrder === TokenBucketPriority.owned) {
          for (const token of sortedErc20TokensWithBalances) {
            if (
              tokenList?.[token.address] ??
              tokenList?.[token.address.toLowerCase()]
            ) {
              const tokenWithTokenListData = buildTokenData({
                ...token,
                ...(token.address
                  ? tokenList?.[token.address] ??
                    tokenList?.[token.address.toLowerCase()]
                  : {}),
              });
              if (tokenWithTokenListData) {
                yield tokenWithTokenListData;
              }
            }
          }
        }

        for (const t of topTokens) {
          const token =
            tokenList?.[t.address] ?? tokenList?.[t.address.toLowerCase()];
          if (token) {
            const tokenWithTokenListData = buildTokenData(token);
            if (tokenWithTokenListData) {
              yield tokenWithTokenListData;
            }
          }
        }

        for (const token of Object.values(tokenList)) {
          const tokenWithTokenListData = buildTokenData(token);
          if (tokenWithTokenListData) {
            yield tokenWithTokenListData;
          }
        }
      })();
    },
    [
      balanceOnActiveChain,
      sortedErc20TokensWithBalances,
      topTokens,
      tokenConversionRates,
      conversionRate,
      currentCurrency,
      chainId,
      tokenList,
    ],
  );

  return filteredTokenListGenerator;
};
