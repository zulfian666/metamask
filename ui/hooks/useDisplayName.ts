import { NameType } from '@metamask/name-controller';
import { useSelector } from 'react-redux';
import { getRemoteTokens } from '../selectors';
import { getNftContractsByAddressOnCurrentChain } from '../selectors/nft';
import { useNames } from './useName';
import { useFirstPartyContractNames } from './useFirstPartyContractName';
import { useNftCollectionsMetadata } from './useNftCollectionsMetadata';

export type UseDisplayNameRequest = {
  value: string;
  preferContractSymbol?: boolean;
  type: NameType;
};

export type UseDisplayNameResponse = {
  name: string | null;
  hasPetname: boolean;
  contractDisplayName?: string;
  image?: string;
};

export function useDisplayNames(
  requests: UseDisplayNameRequest[],
): UseDisplayNameResponse[] {
  const nameRequests = requests.map(({ value, type }) => ({
    value,
    type,
  }));

  const nameEntries = useNames(nameRequests);
  const firstPartyContractNames = useFirstPartyContractNames(nameRequests);
  const nftCollections = useNftCollectionsMetadata(nameRequests);
  const values = requests.map(({ value }) => value);

  const contractInfo = useSelector((state) =>
    // TODO: Replace `any` with type
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (getRemoteTokens as any)(state, values),
  );

  const watchedNftNames = useSelector(getNftContractsByAddressOnCurrentChain);

  return requests.map(({ value, preferContractSymbol }, index) => {
    const nameEntry = nameEntries[index];
    const firstPartyContractName = firstPartyContractNames[index];
    const singleContractInfo = contractInfo[index];
    const watchedNftName = watchedNftNames[value.toLowerCase()]?.name;
    const nftCollectionProperties = nftCollections[value.toLowerCase()];

    let nftCollectionName;
    let nftCollectionImage;

    if (nftCollectionProperties?.isSpam === false) {
      nftCollectionName = nftCollectionProperties?.name;
      nftCollectionImage = nftCollectionProperties?.image;
    }

    const contractDisplayName =
      preferContractSymbol && singleContractInfo?.symbol
        ? singleContractInfo.symbol
        : singleContractInfo?.name;

    const name =
      nameEntry?.name ||
      firstPartyContractName ||
      nftCollectionName ||
      contractDisplayName ||
      watchedNftName ||
      null;

    const hasPetname = Boolean(nameEntry?.name);

    return {
      name,
      hasPetname,
      contractDisplayName,
      image: nftCollectionImage,
    };
  });
}

/**
 * Attempts to resolve the name for the given parameters.
 *
 * @param value - The address or contract address to resolve.
 * @param type - The type of value, e.g. NameType.ETHEREUM_ADDRESS.
 * @param preferContractSymbol - Applies to recognized contracts when no petname is saved:
 * If true the contract symbol (e.g. WBTC) will be used instead of the contract name.
 * @returns An object with two properties:
 * - `name` {string|null} - The display name, if it can be resolved, otherwise null.
 * - `hasPetname` {boolean} - True if there is a petname for the given address.
 */
export function useDisplayName(
  value: string,
  type: NameType,
  preferContractSymbol: boolean = false,
): UseDisplayNameResponse {
  return useDisplayNames([{ preferContractSymbol, type, value }])[0];
}
