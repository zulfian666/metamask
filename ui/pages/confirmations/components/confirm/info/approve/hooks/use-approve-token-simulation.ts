import { TransactionMeta } from '@metamask/transaction-controller';
import { BigNumber } from 'bignumber.js';
import { useSelector } from 'react-redux';
import { getIntlLocale } from '../../../../../../../ducks/locale/locale';
import { useDecodedTransactionData } from '../../hooks/useDecodedTransactionData';
import { useIsNFT } from './use-is-nft';

export const UNLIMITED_MSG = 'UNLIMITED MESSAGE';

const UNLIMITED_THRESHOLD = 10 ** 15;

function isSpendingCapUnlimited(decodedSpendingCap: number) {
  return decodedSpendingCap >= UNLIMITED_THRESHOLD;
}

export const useApproveTokenSimulation = (
  transactionMeta: TransactionMeta,
  decimals: string,
) => {
  const locale = useSelector(getIntlLocale);

  const { isNFT, pending: isNFTPending } = useIsNFT(transactionMeta);

  const decodedResponse = useDecodedTransactionData();

  const { value, pending } = decodedResponse;

  const decodedSpendingCap = value
    ? new BigNumber(value.data[0].params[1].value)
        .dividedBy(new BigNumber(10).pow(Number(decimals)))
        .toNumber()
    : 0;

  const tokenPrefix = isNFT ? '#' : '';
  const formattedTokenNum = isNFT
    ? decodedSpendingCap
    : new Intl.NumberFormat(locale).format(decodedSpendingCap);

  let tokenAmount;
  if (!isNFT && isSpendingCapUnlimited(decodedSpendingCap)) {
    tokenAmount = UNLIMITED_MSG;
  } else {
    tokenAmount = `${tokenPrefix}${formattedTokenNum}`;
  }

  return {
    tokenAmount,
    formattedTokenNum,
    value,
    pending: pending || isNFTPending,
  };
};
