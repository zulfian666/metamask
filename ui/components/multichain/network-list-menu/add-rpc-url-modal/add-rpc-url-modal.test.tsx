import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useI18nContext } from '../../../../hooks/useI18nContext';
import AddRpcUrlModal from './add-rpc-url-modal';

jest.mock('../../../../hooks/useI18nContext', () => ({
  useI18nContext: jest.fn(),
}));

describe('AddRpcUrlModal', () => {
  const useI18nContextMock = useI18nContext as jest.Mock;

  beforeEach(() => {
    useI18nContextMock.mockReturnValue((key: string) => key);
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const { container } = render(<AddRpcUrlModal />);
    expect(container).toMatchSnapshot();
  });

  it('should render the input field with the correct label', () => {
    render(<AddRpcUrlModal />);
    const inputLabel = screen.getByLabelText('additionalRpcUrl');
    expect(inputLabel).toBeInTheDocument();
  });

  it('should render the "Add URL" button with correct text', () => {
    render(<AddRpcUrlModal />);
    const addButton = screen.getByRole('button', { name: 'addUrl' });
    expect(addButton).toBeInTheDocument();
  });

  it('should call the appropriate function when "Add URL" button is clicked', () => {
    const mockAddUrl = jest.fn();
    render(<AddRpcUrlModal />);
    const addButton = screen.getByRole('button', { name: 'addUrl' });
    userEvent.click(addButton);
    expect(mockAddUrl).not.toHaveBeenCalled();
  });
});
