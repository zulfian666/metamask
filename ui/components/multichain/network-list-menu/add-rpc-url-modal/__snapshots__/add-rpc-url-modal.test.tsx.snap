// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRpcUrlModal should render correctly 1`] = `
<div>
  <div
    class="mm-box mm-box--padding-4"
  >
    <div
      class="mm-box mm-form-text-field mm-box--display-flex mm-box--flex-direction-column"
    >
      <label
        class="mm-box mm-text mm-label mm-label--html-for mm-form-text-field__label mm-text--body-sm-medium mm-text--font-weight-medium mm-box--display-inline-flex mm-box--align-items-center mm-box--color-text-default"
        for="additional-rpc-url"
      >
        additionalRpcUrl
      </label>
      <div
        class="mm-box mm-text-field mm-text-field--size-md mm-text-field--truncate mm-form-text-field__text-field mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--align-items-center mm-box--background-color-background-default mm-box--rounded-sm mm-box--border-width-1 box--border-style-solid"
      >
        <input
          autocomplete="off"
          class="mm-box mm-text mm-input mm-input--disable-state-styles mm-text-field__input mm-text--body-md mm-box--margin-0 mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-4 mm-box--color-text-default mm-box--background-color-transparent mm-box--border-style-none"
          focused="false"
          id="additional-rpc-url"
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      class="mm-box mm-text mm-button-base mm-button-base--size-lg mm-button-primary mm-text--body-md-medium mm-box--margin-top-8 mm-box--margin-right-auto mm-box--margin-left-auto mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-4 mm-box--display-block mm-box--justify-content-center mm-box--align-items-center mm-box--width-full mm-box--color-primary-inverse mm-box--background-color-primary-default mm-box--rounded-pill"
    >
      addUrl
    </button>
  </div>
</div>
`;
